Make with_texture_canvas restore target.

czipperz opened the pull request:
The following code will now work:

```
self.canvas.with_texture_canvas(
    &mut self.buffer,
    |canvas| {
        {
            let texture = cache.get(tile);
            if texture.is_some() {
                // this requires the buffer to be the render target
                buffer_render_tile_(canvas, texture.unwrap(), y, x);
                return;
            }
        }

        // cache the new render and then copy it
        let mut texture = gen(creator);
        // this sets the render target
        cache_render_tile(canvas, &mut texture, textures, tile, y, x);
        // this requires the buffer to be the render target
        buffer_render_tile_(canvas, &texture, y, x);
        cache.insert((*tile).clone(), texture);
    }).unwrap();
```

Cobrand commented:
Can you please be more specific about what does this PR fix exactly? Resetting to `ptr::null()` should restore the canvas to the original target... are you saying that this doesn't happen in some cases or something?

I have trouble seeing what is the problem there

czipperz commented:
Say you call `with_texture_canvas`.  Inside that closure I think that the target of the canvas should be the texture.  Inside of an embedded call to `with_texture_canvas`, the canvas should target the texture passed in then.  The problem right now is that the first statement is not always true:

```
canvas.with_texture_canvas(texture1,
|canvas| {
    // here we draw to texture1
    canvas.with_texture_canvas(texture2, ...);
    // here we should draw to texture1.  But without this pull we draw to the Window.
});
```

Cobrand commented:
Oh, what you mean is that it doesn't work properly with nested calls? After the second with_texture_canvas ends, the target is reset to the window, which is unexpected?

I have never seen this use-case and I'd like to know why you have to nest them and not simply make them one after another. Once you confirm that this is a real use-case, I will be fine with merging this PR.

czipperz commented:
Well it doesn't matter if it's a use-case.  Bugs should be fixed.  I have an example straight from the game I'm making at the top.

Here's another example.  It says in the documentation that this method can be expensive.  I have a loop that draws the map:

```
for y in 0..map.height() {
    for x in 0..map.width() {
        render_tile(y, x);
    }
}
```

So every call to render_tile will change the render target to the map texture and draw each tile.  Then every frame, a subsection of the map is presented to the screen.

This loop that prerenders is actually fairly slow if each tile is rendered individually.  So I set up a cache that caches the texture for each different type of tile.  So if the tile isn't cached, a new texture is created and drawn onto.  Then the tile is always cached and can be copied to the map texture.

I don't understand why this library should refuse to make this behavior work when it is the correct behavior for this method.

Cobrand commented:
Don't take it aggressively, I never said I would refuse to make this PR, I just said that I would wait until you told me your use case before merging. If I had merged this directly, you may have called it a day and not answered my question. However it's really important to understand the different use-cases of the users as a maintainer, that's why I wanted to understand what you wanted to do not matter what.

I understand that this is a bug and this will be merged anyway, I just wanted your explanation on that matter.

Speaking of your use case, you will probably want to use with_texture_canvas multiple, which is faster for loops like you've been doing. I don't really get why you need to change the RenderTarget to the map in your case, can't you just change the RenderTarget once and draw a different texture depending on the position?

czipperz commented:
Well in my case the correct thing to do is to refactor the caching out of the rendering loop and then render with the assumption that everything is cached.  This would remove the embedded call to with_texture_canvas.

I am aware of the with_multiple_texture_canvas and do use it at points when appropriate.  However it wouldn't work here.

I forgot to finish my loop example.  Because with_texture_canvas is documented as expensive, I moved that call outside of the loop as to reduce calls.

Refactored code looks like

```
for y { for x { ensure_tile_is_cached(y, x); } }
canvas.with_render_target(map, |canvas| { for y { for x { render_tile(canvas, y, x); } } });
```
